lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -0,0 +1,893 @@
+/// type of the data subfield.
+static FIRRTLType buildBundleType(FIRRTLType dataType, bool isFlip,
+                                  MLIRContext *context) {
+  using BundleElement = std::pair<Identifier, FIRRTLType>;
+  llvm::SmallVector<BundleElement, 3> elements;
+
+  // Add valid and ready subfield to the bundle.
+  Identifier validId = Identifier::get("valid", context);
+  Identifier readyId = Identifier::get("ready", context);
+  UIntType signalType = UIntType::get(context, 1);
+  if (isFlip) {
+    elements.push_back(BundleElement(validId, FlipType::get(signalType)));
+    elements.push_back(BundleElement(readyId, signalType));
+  } else {
+    elements.push_back(BundleElement(validId, signalType));
+    elements.push_back(BundleElement(readyId, FlipType::get(signalType)));
+  }
+
+  // Add data subfield to the bundle if dataType is not a null.
+  if (dataType) {
+    Identifier dataId = Identifier::get("data", context);
+    if (isFlip)
+      elements.push_back(BundleElement(dataId, FlipType::get(dataType)));
+    else
+      elements.push_back(BundleElement(dataId, dataType));
+  }
+
+  return BundleType::get(ArrayRef<BundleElement>(elements), context);
+}
+
+/// Return a FIRRTL bundle type (with data, valid, and ready subfields) given a
+/// standard data type. Current supported data types are integer (signed,
+/// unsigned, and signless), index, and none.
+static FIRRTLType getBundleType(Type type, bool isFlip) {
+  // If the input is already converted to a bundle type elsewhere, itself will
+  // be returned after cast.
+  if (auto bundleType = type.dyn_cast<BundleType>())
+    return bundleType;
+
+  MLIRContext *context = type.getContext();
+  switch (type.getKind()) {
+  case StandardTypes::Integer: {
+    IntegerType integerType = type.cast<IntegerType>();
+    int width = integerType.getWidth();
+
+    switch (integerType.getSignedness()) {
+    case IntegerType::Signed:
+      return buildBundleType(SIntType::get(context, width), isFlip, context);
+    case IntegerType::Unsigned:
+      return buildBundleType(UIntType::get(context, width), isFlip, context);
+    // ISSUE: How to handle signless integers? Should we use the AsSIntPrimOp
+    // or AsUIntPrimOp to convert?
+    case IntegerType::Signless:
+      return buildBundleType(SIntType::get(context, width), isFlip, context);
+    }
+  }
+  // Currently we consider index type as 64-bits signed integer.
+  case StandardTypes::Index: {
+    int width = type.cast<IndexType>().kInternalStorageBitWidth;
+    return buildBundleType(SIntType::get(context, width), isFlip, context);
+  }
+  case StandardTypes::None:
+    return buildBundleType(/*dataType=*/nullptr, isFlip, context);
+  default:
+    assert(false && "Unsupported data type. Supported data types: integer "
+                    "(signed, unsigned, signless), index, none");
+    return FIRRTLType(nullptr);
+  }
+}
+
+static Value createConstantOp(Location insertLoc, FIRRTLType opType, int value,
```

Comment:
is "int value" really enough here?  What about wider things?

Reply:
Specifically, it looks like buildConstantLogic will implicitly narrow an APInt to an int, then this will widen it out, losing data.